README.md
# NAME

Net::BitTorrent::DHT - Kademlia-like DHT Node

# Synopsis

    use Net::BitTorrent::DHT;
    use AnyEvent;
    use Bit::Vector;
    # Standalone node with user-defined port and boot_nodes
    my $dht = Net::BitTorrent::DHT->new(
          port => [1337 .. 1340, 0],
          boot_nodes =>
              [['router.bittorrent.com', 6881], ['router.utorrent.com', 6881]]
    );

    my $peer_quest
    = $dht->get_peers(Bit::Vector->new_Hex('ab97a7bca78f2628380e6609a8241a7fb02aa981'), \&dht_cb);

    # tick, tick, tick, ...
    AnyEvent->condvar->recv;

    sub dht_cb {
        my ($infohash, $node, $peers) = @_;
        printf "We found %d peers for %s from %s:%d via DHT\n\t%s\n",
            scalar(@$peers),
            $infohash->to_Hex, $node->host, $node->port,
            join ', ', map { sprintf '%s:%d', @$_ } @$peers;
    }

# Description

BitTorrent uses a "distributed sloppy hash table" (DHT) for storing peer
contact information for "trackerless" torrents. In effect, each peer becomes a
tracker. The protocol is based on [Kademila](#kademlia) and is implemented
over UDP.

# Methods

[Net::BitTorrent::DHT](https://metacpan.org/pod/Net::BitTorrent::DHT)'s API is simple but powerful.
...well, I think so anyway.

# Net::BitTorrent::DHT->new( )

The constructor accepts a number different arguments which all greatly affect
the function of your DHT node. Any combination of the following arguments may
be used during construction.

Note that [standalone](https://metacpan.org/pod/Net::BitTorrent::DHT::Standalone) DHT nodes do not
support or require the `client` argument but internally a
[Net::BitTorrent](https://metacpan.org/pod/Net::BitTorrent) client is passed and serves as the parent
of this node. For brevity, the following examples assume you are building a
[standalone node](https://metacpan.org/pod/Net::BitTorrent::DHT::Standalone) (for reasearch, etc.).

## Net::BitTorrent::DHT->new( nodeid => ... )

During construction, our local DHT nodeID can be set during construction. This
is mostly useful when creating a
[standalone DHT node](https://metacpan.org/pod/Net::BitTorrent::DHT::Standalone).

    use Net::BitTorrent::DHT;
    # Bit::Vector object
    use Bit::Vector;
    my $node_c = Net::BitTorrent::DHT->new(
        nodeid => Bit::Vector->new_Hex( 160, 'ABCD' x 10 )
    );
    # A SHA1 digest
    use Digest::SHA;
    my $node_d = Net::BitTorrent::DHT->new(
            nodeid => Bit::Vector->new_Hex( 160, Digest::SHA::sha1( $possibly_random_value ) )
    );

Note that storing and reusing DHT nodeIDs over a number of sessions may seem
advantagious (as if you had a "reserved parking place" in the DHT network) but
will likely not improve performance as unseen nodeIDs are removed from remote
routing tables after a half hour.

NodeIDs, are 160-bit integers.

## Net::BitTorrent::DHT->new( port => ... )

Opens a specific UDP port number to the outside world on both IPv4 and IPv6.

    use Net::BitTorrent::DHT;
    # A single possible port
    my $node_a = Net::BitTorrent::DHT->new( port => 1123 );
    # A list of ports
    my $node_b = Net::BitTorrent::DHT->new( port => [1235 .. 9875] );

Note that when handed a list of ports, they are each tried until we are able
to bind to the specific port.

# Net::BitTorrent::DHT->find\_node( $target, $callback )

This method asks for remote nodes with nodeIDs closer to our target. As the
remote nodes respond, the callback is called with the following arguments:

- target

    This is the target nodeid. This is useful when you've set the same callback
    for multiple, concurrent `find_node( )` [quest](#quests-and-callbacks).

    Targets are 160-bit [Bit::Vector](https://metacpan.org/pod/Bit::Vector) objects.

- node

    This is a blessed object. TODO.

- nodes

    This is a list of ip:port combinations the remote node claims are close to our
    target.

A single `find_node` [quest](https://metacpan.org/pod/Net::BitTorrent::Notes#Quests-and-Callbacks)
is an array ref which contains the following data:

- target

    This is the target nodeID.

- coderef

    This is the callback triggered as we locate new peers.

- nodes

    This is a list of nodes we have announced to so far.

- timer

    This is an [AnyEvent](https://metacpan.org/pod/AnyEvent) timer which is triggered every few minutes.

    Don't modify this.

# Net::BitTorrent::DHT->get\_peers( $infohash, $callback )

This method initiates a search for peers serving a torrent with this infohash.
As they are found, the callback is called with the following arguments:

- infohash

    This is the infohash related to these peers. This is useful when you've set
    the same callback for multiple, concurrent `get_peers( )` quests. This is a
    160-bit [Bit::Vector](https://metacpan.org/pod/Bit::Vector) object.

- node

    This is a blessed object. TODO.

- peers

    This is an array ref of peers sent to us by aforementioned remote node.

A single `get_peers` [quest](https://metacpan.org/pod/Net::BitTorrent::Notes#Quests-and-Callbacks)
is an array ref which contains the following data:

- infohash

    This is the infohash related to these peers. This is a 160-bit
    [Bit::Vector](https://metacpan.org/pod/Bit::Vector) object.

- coderef

    This is the callback triggered as we locate new peers.

- peers

    This is a compacted list of all peers found so far. This is probably more
    useful than the list passed to the callback.

- timer

    This is an [AnyEvent](https://metacpan.org/pod/AnyEvent) timer which is triggered every five minutes.
    When triggered, the node requests new peers from nodes in the bucket nearest
    to the infohash.

    Don't modify this.

# Net::BitTorrent::DHT->**announce\_peer**( $infohash, $port, $callback )

This method announces that the peer controlling the querying node is
downloading a torrent on a port. These outgoing queries are sent to nodes
'close' to the target infohash. As the remote nodes respond, the callback is
called with the following arguments:

- infohash

    This is the infohash related to this announcment. This is useful when you've
    set the same callback for multiple, concurrent `announce_peer( )`
    [quest](#quests-and-callbacks). Infohashes are 160-bit
    [Bit::Vector](https://metacpan.org/pod/Bit::Vector) objects.

- port

    This is port you defined above.

- node

    This is a blessed object. TODO.

A single `announce_peer` [quest](#quests-and-callbacks) is an array ref
which contains the following data:

- infohash

    This is the infohash related to these peers. This is a 160-bit
    [Bit::Vector](https://metacpan.org/pod/Bit::Vector) object.

- coderef

    This is the callback triggered as we locate new peers.

- port

    This is port you defined above.

- nodes

    This is a list of nodes we have announced to so far.

- timer

    This is an [AnyEvent](https://metacpan.org/pod/AnyEvent) timer which is triggered every few minutes.

    Don't modify this.

`announce_peer` queries require a token sent in reply to a `get_peers` query
so they should be used together.

    use Net::BitTorrent::DHT;
    my $node = Net::BitTorrent::DHT->new( );
    my $quest_a = $dht->announce_peer(Bit::Vector->new_Hex('A' x 40), 6881, \&dht_cb);
    my $quest_b = $dht->announce_peer(Bit::Vector->new_Hex('1' x 40), 9585, \&dht_cb);

    sub dht_cb {
        my ($infohash, $port, $node) = @_;
        say sprintf '%s:%d now knows we are serving %s on port %d',
            $node->host, $node->port, $infohash->to_Hex, $port;
    }

# Net::BitTorrent::DHT->dump\_ipv4\_buckets( )

This is a quick utility method which returns or prints (depending on context)
a list of the IPv4-based routing table's bucket structure.

    use Net::BitTorrent::DHT;
    my $node = Net::BitTorrent::DHT->new( );
    # After some time has passed...
    $node->dump_ipv4_buckets; # prints to STDOUT with say
    my @dump = $node->dump_ipv4_buckets; # returns list of lines

# Net::BitTorrent::DHT->dump\_ipv6\_buckets( )

This is a quick utility method which returns or prints (depending on context)
a list of the IPv6-based routing table's bucket structure.

    use Net::BitTorrent::DHT;
    my $node = Net::BitTorrent::DHT->new( );
    # After some time has passed...
    $node->dump_ipv6_buckets; # prints to STDOUT with say
    my @dump = $node->dump_ipv6_buckets; # returns list of lines

# Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

# License and Legal

Copyright (C) 2008-2014 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
[The Artistic License 2.0](http://www.perlfoundation.org/artistic_license_2_0).
See the `LICENSE` file included with this distribution or
[notes on the Artistic License 2.0](http://www.perlfoundation.org/artistic_2_0_notes)
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
[Creative Commons Attribution-Share Alike 3.0 License](http://creativecommons.org/licenses/by-sa/3.0/us/legalcode).
See the
[clarification of the CCA-SA3.0](http://creativecommons.org/licenses/by-sa/3.0/us/).

Neither this module nor the [Author](#author) is affiliated with BitTorrent,
Inc.


Changes
Revision history for Perl extension Net-BitTorrent-DHT

{{$NEXT}}

    - Declare Type::Standard dependency
    
v1.0.2 2014-06-26T17:00:33Z

    - Serve as a standalone node by default
    - Update to NB::Protocol v1.0.2 and above

v1.0.1 2014-06-21T19:57:48Z

    - Generate local node id based on external IP (work in progress)
    - Use a condvar in address resolver instead of forcing the event loop

v1.0.0 2014-06-21T16:04:36Z

    - original version (broken from unstable Net::BitTorrent dist)


lib/Net/BitTorrent/DHT/Node.pm
package Net::BitTorrent::DHT::Node;
use Moose;
use Moose::Util::TypeConstraints;
use AnyEvent;
use Net::BitTorrent::Protocol qw[:dht :compact];
use feature qw[state];
use Scalar::Util;
use Types::Standard qw[Bool Int Ref Str];
our $VERSION = 'v1.0.2';
eval $VERSION;
#
sub BUILD {1}

#
has port     => (isa => Int, is => 'ro', required => 1);
has host     => (isa => Str, is => 'ro', required => 1);
has sockaddr => (isa => Str, is => 'ro', required => 1, lazy_build => 1);

sub _build_sockaddr {
    Net::BitTorrent::DHT::sockaddr($_[0]->host, $_[0]->port);
}
has ipv6 => (isa => Bool, is => 'ro', lazy_build => 1);
sub _build_ipv6 { length shift->sockaddr == 28 }
for my $dir (qw[in out]) {
    has 'announce_peer_token_'
        . $dir => (isa     => 'HashRef[Str]',
                   is      => 'ro',
                   traits  => ['Hash'],
                   handles => {'_set_announce_peer_token_' . $dir => 'set',
                               '_get_announce_peer_token_' . $dir => 'get',
                               '_del_announce_peer_token_' . $dir => 'delete',
                               'has_announce_peer_token_' . $dir  => 'defined'
                   },
                   default => sub { {} }
        );
}
has v => (isa => Str, is => 'ro', writer => '_v', predicate => 'has_v');
has bucket => (isa       => 'Net::BitTorrent::DHT::Bucket',
                 is        => 'ro',
                 writer    => 'assign_bucket',
                 weak_ref  => 1,
                 predicate => 'has_bucket'
);
has routing_table => (isa        => 'Net::BitTorrent::DHT::RoutingTable',
                        is         => 'ro',
                        predicate  => 'has_routing_table',
                        writer     => '_routing_table',
                        weak_ref   => 1,
                        lazy_build => 1,
                        handles    => [qw[send dht tracker]]
);
around send => sub {
    my ($code, $self, $packet, $reply) = @_;
    $code->($self, $self, $packet, !!$reply);
};
has nodeid => (isa       => 'Bit::Vector',
                 is        => 'ro',
                 writer    => '_set_nodeid',
                 predicate => 'has_nodeid'
);
after _set_nodeid => sub {
    $_[0]->routing_table->assign_node($_[0]);
    $_[0]->routing_table->del_node($_[0]) if !$_[0]->has_bucket;
};
has outstanding_requests => (isa     => 'HashRef[HashRef]',
                               is      => 'ro',
                               traits  => ['Hash'],
                               handles => {add_request    => 'set',
                                           get_request    => 'get',
                                           del_request    => 'delete',
                                           expire_request => 'delete',
                                           is_expecting   => 'defined'
                               },
                               init_arg => undef,
                               default  => sub { {} }
);
after expire_request => sub { shift->inc_fail };
around add_request => sub {
    my ($code, $self, $tid, $args) = @_;
    Scalar::Util::weaken $self;
    $args->{'timeout'} //= AE::timer(
        20, 0,
        sub {
            $self->expire_request($tid) if $self;    # May ((poof)) $self
        }
    );
    $code->($self, $tid, $args);
};
has ping_timer => (
         isa      => Ref,                               # ArrayRef|EV::Timer
         builder  => '_build_ping_timer',
         is       => 'ro',
         init_arg => undef,
         writer   => '_ping_timer'
);

sub _build_ping_timer {
    my ($self) = @_;
    Scalar::Util::weaken $self;
    AE::timer(60 * 10, 60 * 10, sub { $self->ping if $self });
}
has seen => (
    isa        => Int,
    is         => 'ro',
    lazy_build => 1,
    init_arg   => undef,
    writer     => '_set_seen',
    handles    => {
        touch  => sub { shift->_set_seen(time) },
        active => sub {
            return time - shift->seen <= 15 * 60;
            }
    }
);
for my $type (qw[get_peers find_node announce_peer]) {
    has 'prev_'
        . $type => (isa        => 'HashRef[Int]',
                    is         => 'ro',
                    lazy_build => 1,
                    builder    => '_build_prev_X',
                    init_arg   => undef,
                    traits     => ['Hash'],
                    handles    => {
                                'get_prev_' . $type     => 'get',
                                'set_prev_' . $type     => 'set',
                                'defined_prev_' . $type => 'defined'
                    }
        );
}
sub _build_prev_X { {} }
after BUILD => sub {
    my ($self) = @_;
    Scalar::Util::weaken $self;
    $self->_ping_timer(AE::timer(rand(30), 0, sub { $self->ping }));
};
has birth => (is         => 'ro',
                isa        => Int,
                init_arg   => undef,
                lazy_build => 1
);
sub _build_birth {time}

sub ping {
    my ($self) = @_;
    state $tid = 'a';
    my $packet = build_ping_query('p_' . $tid,
                                      pack('H*', $self->dht->nodeid->to_Hex));
    my $sent = $self->send($packet);
    return $self->inc_fail() if !$sent;
    $self->add_request('p_' . $tid, {type => 'ping'});
    $tid++;
}

sub _reply_ping {
    my ($self, $tid) = @_;
    my $packet
        = build_ping_reply($tid, pack('H*', $self->dht->nodeid->to_Hex));
    my $sent = $self->send($packet, 1);
    $self->inc_fail() if !$sent;
    return $sent;
}

sub find_node {
    my ($self, $target) = @_;
    return
        if $self->defined_prev_find_node($target->to_Hex)
        && $self->get_prev_find_node($target->to_Hex) > time - (60 * 15);
    state $tid = 'a';
    my $packet =
        build_find_node_query('fn_' . $tid,
                                  pack('H*', $self->dht->nodeid->to_Hex),
                                  pack('H*', $target->to_Hex));
    my $sent = $self->send($packet);
    return $self->inc_fail() if !$sent;
    $self->add_request('fn_' . $tid,
                       {type => 'find_node', target => $target});
    $tid++;
    $self->set_prev_find_node($target->to_Hex, time);
}

sub _reply_find_node {
    my ($self, $tid, $target) = @_;
    my $nodes
        = compact_ipv4(
                 map { [$_->host, $_->port] }
                     @{$self->routing_table->nearest_bucket($target)->nodes});
    return if !$nodes;
    my $packet = build_find_node_reply($tid, pack('H*', $target->to_Hex),
                                           $nodes);
    my $sent = $self->send($packet, 1);
    $self->inc_fail() if !$sent;
    return $sent;
}

sub get_peers {
    my ($self, $info_hash) = @_;
    return
        if $self->defined_prev_get_peers($info_hash->to_Hex)
        && $self->get_prev_get_peers($info_hash->to_Hex) > time - (60 * 15);
    state $tid = 'a';
    my $packet =
        build_get_peers_query('gp_' . $tid,
                                  pack('H*', $self->dht->nodeid->to_Hex),
                                  pack('H*', $info_hash->to_Hex)
        );
    my $sent = $self->send($packet);
    return $self->inc_fail() if !$sent;
    $self->add_request('gp_' . $tid,
                       {type => 'get_peers', info_hash => $info_hash});
    $tid++;
    $self->set_prev_get_peers($info_hash->to_Hex, time);
}

sub _reply_get_peers {
    my ($self, $tid, $id) = @_;
    if (!$self->has_announce_peer_token_out($id->to_Hex)) {
        state $announce_peer_token = 'aa';
        $announce_peer_token = 'aa' if length $announce_peer_token == 3;
        $self->_set_announce_peer_token_out($id->to_Hex,
                                            $announce_peer_token++);
    }
    my $nodes
        = compact_ipv4(
                     map { [$_->host, $_->port] }
                         @{$self->routing_table->nearest_bucket($id)->nodes});
    my @values = grep { defined $_ } map {
        Net::BitTorrent::Protocol::BEP23::Compact::compact_ipv4(
                                                           [$_->[0], $_->[1]])
    } @{$self->tracker->get_peers($id) || []};
    return if (!@values && !$nodes);
    my $packet =
        build_get_peers_reply($tid,
                                  $id->to_Hex,
                                  \@values,
                                  $nodes,
                                  $self->_get_announce_peer_token_out(
                                                                   $id->to_Hex
                                  )
        );
    my $sent = $self->send($packet, 1);
    $self->inc_fail() if !$sent;
    return $sent;
}

sub announce_peer {
    my ($self, $info_hash, $port) = @_;
    return
        if $self->defined_prev_announce_peer($info_hash->to_Hex)
        && $self->get_prev_announce_peer($info_hash->to_Hex)
        > time - (60 * 30);
    return if !$self->has_announce_peer_token_in($info_hash->to_Hex);
    state $tid = 'a';
    my $packet =
        build_announce_peer_query('an_' . $tid,
                                      pack('H*', $self->dht->nodeid->to_Hex),
                                      pack('H*', $info_hash->to_Hex),
                                      $self->_get_announce_peer_token_in(
                                                            $info_hash->to_Hex
                                      ),
                                      $port
        );
    my $sent = $self->send($packet);
    return $self->inc_fail() if !$sent;
    $self->add_request('an_' . $tid,
                       {type => 'announce_peer', info_hash => $info_hash});
    $tid++;
    $self->set_prev_announce_peer($info_hash->to_Hex, time);
}

sub _reply_announce_peer {
    my ($self, $tid, $info_hash, $a_ref) = @_;
    my $packet;
    if ((!$self->has_announce_peer_token_out($info_hash->to_Hex))
        || ($self->_get_announce_peer_token_out($info_hash->to_Hex) ne
            $a_ref->{'token'})
        )
    {   $packet =
            build_error_reply($tid,
                                  [203,
                                   'Incorrect write token in announce_peer'
                                  ]
            );
    }
    elsif (
         !$self->tracker->add_peer($info_hash, [$self->host, $a_ref->{'port'}]
         )
        )
    {   $packet = build_error_reply($tid,
                                      [202, 'Failed to add peer to tracker']);
    }
    else {
        $packet = build_announce_peer_reply($tid,
                                      pack('H*', $self->dht->nodeid->to_Hex));
    }
    my $sent = $self->send($packet, 1);
    $self->inc_fail() if !$sent;
    return $sent;
}
has fail => (
    isa      => Int,
    traits   => ['Counter'],
    default  => 0,
    is       => 'ro',
    handles  => {inc_fail => 'inc'},
    init_arg => undef,
    trigger  => sub {
        my ($self, $new, $old) = @_;
        $self->routing_table->del_node($self)
            if $new == ($self->has_bucket ? 5 : 1);
    }
);
1;

=pod

=head1 NAME

Net::BitTorrent::DHT::Node - A single node in a DHT routing table

=head1 Description

TODO

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2008-2014 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

Neither this module nor the L<Author|/Author> is affiliated with BitTorrent,
Inc.

=cut


lib/Net/BitTorrent/DHT/Bucket.pm
package Net::BitTorrent::DHT::Bucket;
use Moose;
use Moose::Util::TypeConstraints;
use AnyEvent;
use Bit::Vector;
use Scalar::Util;
use feature qw[state];
our $VERSION = 'v1.0.2';
eval $VERSION;

# Stub
sub BUILD {1}
#
my $K = 8;    # max nodes per-bucket
#
has _id => (isa => 'Str', is => 'ro', lazy_build => 1);
sub _build__id { state $id = 'a'; $id++ }
has floor => (isa        => 'Bit::Vector',
                is         => 'ro',
                lazy_build => 1
);
sub _build_floor { Bit::Vector->new(160) }
has next => (isa       => 'Net::BitTorrent::DHT::Bucket',
               is        => 'ro',
               writer    => '_next',
               predicate => 'has_next',
               weak_ref  => 1
);
for my $type ('', 'backup_') {
    has $type . 'nodes' => (
        isa => subtype(
            as 'ArrayRef[Net::BitTorrent::DHT::Node]' =>
                where { scalar @$_ <= $K } => message {
                sprintf 'Too many %s nodes! %d with max %d', $type,
                    scalar @$_, $K;
            }
        ),
        is      => 'ro',
        default => sub { [] },
        traits  => ['Array'],
        handles => {'pop_' . $type . 'nodes'     => 'pop',
                    'add_' . $type . 'node'      => 'push',
                    'shift_' . $type . 'nodes'   => 'shift',
                    'unshift_' . $type . 'nodes' => 'unshift',
                    'splice_' . $type . 'nodes'  => 'splice',
                    'count_' . $type . 'nodes'   => 'count',
                    'grep_' . $type . 'nodes'    => 'grep',
                    'map_' . $type . 'nodes'     => 'map',
                    'first_' . $type . 'node'    => 'first',
                    'clear_' . $type . 'nodes'   => 'clear'
        }
    );
}
around add_node => sub {
    my ($code, $self, $node) = @_;
    if ($self->count_nodes == $K) {
        if ($self->_id eq
            $self->routing_table->nearest_bucket($self->dht->nodeid)->_id)
        {   return $self->add_node($node) if $self->split();
        }
        return $self->add_backup_node($node);
    }
    return if $self->grep_nodes(
        sub {
            $_->nodeid->Lexicompare($node->nodeid) == 0;
        }
    );
    $code->($self, $node);
    return $node->assign_bucket($self);
};
around add_backup_node => sub {
    my ($code, $self, $node) = @_;
    return if $self->count_backup_nodes == $K;
    return
        if $self->grep_backup_nodes(
        sub {
            $_->nodeid->Lexicompare($node->nodeid) == 0;
        }
        );
    return $code->($self, $node);
};
has routing_table => (isa      => 'Net::BitTorrent::DHT::RoutingTable',
                        is       => 'ro',
                        required => 1,
                        weak_ref => 1,
                        handles  => [qw[dht]]
);
has last_changed => (isa => 'Int', is => 'ro', default => time);
has clearing_house =>
    (isa => 'Ref', is => 'ro', writer => '_build_clearing_house');
after BUILD => sub {
    my $self = shift;
    $self->_build_clearing_house(
        AE::timer(
            30 * 60,
            30 * 60,
            sub {
                Scalar::Util::weaken($self)
                    if !Scalar::Util::isweak($self);
                $self->routing_table->del_node($_)
                    for $self->grep_nodes(sub { $_->fail && !$_->seen });
            }
        )
    );
};
has find_node_quest => (isa        => 'ArrayRef|Maybe',
                          is         => 'ro',
                          init_arg   => undef,
                          lazy_build => 1,
                          writer     => '_find_node_quest',
                          clearer    => '_reset_find_node_quest'
);

sub _build_find_node_quest {
    my ($self) = @_;
    $self->dht->find_node(
        $self->middle,
        sub {
            Scalar::Util::weaken($self) if !Scalar::Util::isweak($self);
            $self->_find_node_quest(undef)
                if $self->count_backup_nodes == $K;
        }
    );
}
after BUILD => sub { $_[0]->find_node_quest() };
after BUILD => sub { };
#
sub ceil {
    my $self = shift;
    my $ceil = Bit::Vector->new_Hex(160, 'F' x 40);
    if ($self->has_next) {
        $ceil->Copy($self->next->floor);
        $ceil->add($ceil, Bit::Vector->new_Dec(160, 1), 0);
    }
    return $ceil;
}

sub middle {
    my $self  = shift;
    my $floor = Bit::Vector->new(160);
    $floor->Copy($self->floor);
    my $range = Bit::Vector->new(160);
    $range->subtract($self->ceil, $floor, 0);
    my $new_floor = Bit::Vector->new(160);
    {    # Resize for overflow
        $_->Resize(161) for $range, $floor, $new_floor;
        my $half_range = Bit::Vector->new(161);
        $half_range->Divide($range,
                            Bit::Vector->new_Dec(161, 2),
                            Bit::Vector->new(161));
        $new_floor->add($floor, $half_range, 0);
        $_->Resize(160) for $new_floor;
    }
    return $new_floor;
}

sub split {
    my ($self) = @_;
    my $new_floor = $self->middle;
    my $new_bucket =
        Net::BitTorrent::DHT::Bucket->new(
                                routing_table => $self->routing_table,
                                ($self->has_next ? (next => $self->next) : ())
        );
    $new_bucket->floor->Copy($new_floor);
    $self->_next($new_bucket);
    $self->routing_table->add_bucket($new_bucket);
    {
        my @nodes = (@{$self->nodes}, @{$self->backup_nodes});
        $self->clear_nodes;
        $self->clear_backup_nodes;
        $self->routing_table->assign_node($_) for @nodes;
    }
    return 1;
}

sub _del_node {
    my ($self, $node) = @_;
    for my $i (0 .. $self->count_nodes) {
        if ($node->nodeid->Lexicompare($self->nodes->[$i]->nodeid) == 0) {
            $self->splice_nodes($i, 1, ());    # Post 'Help Wanted' sign
            last;
        }
    }
    $self->add_node($self->shift_backup_nodes)
        if $self->count_backup_nodes;          # Take sign down
    $self->_reset_find_node_quest if $self->count_nodes < $K;
}
1;

=pod

=head1 NAME

Net::BitTorrent::DHT::Bucket - A single bucket of nodes in a DHT routing table

=head1 Description

TODO

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2008-2014 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

Neither this module nor the L<Author|/Author> is affiliated with BitTorrent,
Inc.

=cut


lib/Net/BitTorrent/DHT/Tracker.pm
package Net::BitTorrent::DHT::Tracker;
use Moose;
our $VERSION = 'v1.0.2';
eval $VERSION;
#
has peers => (isa      => 'HashRef[ArrayRef]',
                is       => 'ro',
                default  => sub { {} },
                init_arg => undef,
                traits   => ['Hash'],
                handles  => {
                            get_peers  => 'get',
                            _set_peers => 'set',
                            has_peers  => 'defined',
                            del_peers  => 'delete'
                }
);
around get_peers => sub {
    my ($code, $self, $infohash) = @_;
    $code->($self, blessed $infohash ? $infohash->to_Hex : $infohash);
};

sub add_peer {
    my ($self, $infohash, $peer) = @_;
    return $self->_set_peers($infohash->to_Hex, [$peer])
        if !$self->has_peers($infohash->to_Hex);
    push(@{$self->get_peers($infohash->to_Hex)}, $peer)
        if !grep { $_->[0] eq $peer->[0] && $_->[1] eq $peer->[1] }
        @{$self->get_peers($infohash->to_Hex)};
}
has routing_table => (isa      => 'Net::BitTorrent::DHT::RoutingTable',
                        is       => 'ro',
                        required => 1,
                        weak_ref => 1,
                        handles  => [qw[dht]]
);
1;

=pod

=head1 NAME

Net::BitTorrent::DHT::Tracker - Psudo-tracker for the DHT

=head1 Description

Nothing to see here.

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2008-2014 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

Neither this module nor the L<Author|/Author> is affiliated with BitTorrent,
Inc.

=cut


eg/standalone_dht.pl
use strict;
use warnings;
use AnyEvent;
use lib '../lib';
use Net::BitTorrent::DHT;
use Bit::Vector;
$|++;
our $VERSION = 'v1.0.0';
eval $VERSION;

# Standalone node with user-defined port and boot_nodes
my $dht = Net::BitTorrent::DHT->new(
          port => [1337 .. 1340, 0],
          boot_nodes =>
              [['router.bittorrent.com', 6881], ['router.utorrent.com', 6881]]
);

# Two 'quests' for peers (these are two popular Ubuntu swarms)
# See http://torrent.ubuntu.com:6969/ for updated list
my $quest_A
    = $dht->get_peers(Bit::Vector->new_Hex(160, 'ab97a7bca78f2628380e6609a8241a7fb02aa981'), \&dht_cb);
my $quest_B
    = $dht->get_peers(Bit::Vector->new_Hex(160, '79053908d904e26b45e5b63a6d9b121b59ed7375'), \&dht_cb);

# Let's stay up to date with what's going on in the routing table
my $timer = AE::timer 60 * 2, 60 * 5,
    sub { $dht->dump_ipv4_buckets; $dht->dump_ipv6_buckets };
END { $dht->dump_ipv4_buckets && $dht->dump_ipv6_buckets if $dht }

# tick, tick, tick, ...
AnyEvent->condvar->recv;

sub dht_cb {
    my ($infohash, $node, $peers) = @_;
    printf "We found %d peers for %s from %s:%d via DHT\n\t%s\n",
        scalar(@$peers),
        $infohash->to_Hex, $node->host, $node->port,
        join ', ', map { sprintf '%s:%d', @$_ } @$peers;
}

=pod

=head1 NAME

standalone_dht - Quick/dirty demo of a standalone DHT node

=head1 Description

This is just perfect for research. I'll breakdown what's going on in a later
version

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2008-2014 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

Neither this module nor the L<Author|/Author> is affiliated with BitTorrent,
Inc.

=cut
